cloud-init/qemu_performance.py

```diff
@@ -0,0 +1,549 @@
+        ]
+    else:
+        raise RuntimeError(
+            f"Invalid deb_path provided: {deb_path}. Expected local .deb or" " ppa:"
+        )
+    for command in commands:
+        logging.debug(f"--- Running: {' '.join(command)}")
+        subprocess.check_call(command)
+    return new_img_path
+
+
+def collect_bootspeed_data(instance):
+    """Collect and process bootspeed data from the instance."""
+    start_time = time.time()
+    data = {"time_at_ssh": retry_cmd(instance, "date --utc +'%b %d %H:%M:%S.%N'")}
+    ssh_time = time.time()
+    data["cloudinit_status"] = json.loads(
+        instance.execute("cloud-init status --format=json --wait")
+    )
+    cloudinit_done_time = time.time()
+    data["client_time_to_ssh"] = ssh_time - start_time
+    data["client_time_to_cloudinit_done"] = cloudinit_done_time - start_time
+    data["image_builddate"] = retry_cmd(
+        instance, "grep serial /etc/cloud/build.info"
+    ).split()[-1]
+    data["systemd_analyze"] = retry_cmd(instance, "systemd-analyze")
+    data["systemd_analyze_blame"] = retry_cmd(instance, "systemd-analyze blame")
+    data["cloudinit_version"]: retry_cmd(instance, "dpkg-query -W cloud-init").split()[
+        1
+    ]
+    data["cloudinit_analyze"] = instance.execute("cloud-init analyze show")
+    data["cloudinit_analyze_blame"] = instance.execute("cloud-init analyze blame")
+
+    for service in (
+        "cloud-init.service",
+        "cloud-init-local.sevice",
+        "cloud-config.service",
+        "cloud-final.service",
+        "systemd-user-sessions.service",  # Affects sytem login time
+        "snapd.seeded.service",  # Has impact on cloud-config.service
+    ):
+        data[
+            f"criticalchain_{service.replace('-', '_').replace('.', '_')}"
+        ] = instance.execute(f"systemd-analyze critical-chain {service}")
+    data["file_journalctl.log"] = instance.execute(
+        "journalctl -o short-precise"
+    )
+    data["file_cloud_init.log"] = instance.execute(
+        "cat /var/log/cloud-init.log"
+    )
+    return data
+
+
+def get_max_min_avg(samples: list):
+    return {
+        "max": max(samples),
+        "min": min(samples),
+        "avg": float(sum(samples) / len(samples)),
+        "stdev": 0.0 if len(samples) < 2 else statistics.stdev(samples),
+    }
+
+
+def process_bootspeed_data(boot_sample: dict):
+    """Process semi-structured boot_sample, extracting boot times
+
+    Process systemd-analyze blame and cloud-init analyze blame output
+    to create top-level keys and associated boot time cost in seconds
+    for the following boot data:
+       time_systemd_kernel, time_systemd_userspace,
+       time_systemd_blames (dict of services and costs) and
+       time_cloudnit_blames (dict of config modules, boot stages).
+    """
+    analyze_times = re.match(
+        r".*in (?P<kernel>[^s]+)s \(kernel\) \+ (?P<user>[^s]+)s \(userspace\)",
+        boot_sample["systemd_analyze"],
+    )
+    if analyze_times:
```

Comment:
If this conditional is false, this causes an exception to throw:

```
Traceback (most recent call last):
  File "/usr/lib/python3.12/pdb.py", line 1960, in main
    pdb._run(target)
  File "/usr/lib/python3.12/pdb.py", line 1754, in _run
    self.run(target.code)
  File "/usr/lib/python3.12/bdb.py", line 606, in run
    exec(cmd, globals, locals)
  File "<string>", line 1, in <module>
  File "/home/holmanb/git/server-test-scripts/cloud-init/qemu_performance.py", line 534, in <module>
    inspect_boot_time_deltas_for_deb(
  File "/home/holmanb/git/server-test-scripts/cloud-init/qemu_performance.py", line 435, in inspect_boot_time_deltas_for_deb
    orig_avg = update_averages(orig_data, data_dir, "orig")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/holmanb/git/server-test-scripts/cloud-init/qemu_performance.py", line 245, in update_averages
    avg_data[k]["samples"].append(boot_sample[k])
                                  ~~~~~~~~~~~^^^
KeyError: 'time_systemd_kernel'
```

More robust handling would be preferred.